File: framework/languages/dotnet/samples/SupportChat/Server/Support/Infrastructure/ZLink/Spots/EntrySpot/Handlers/JoinConversationActorHandler.cs
using SupportChat.Server.Configuration;
using SupportChat.Server.Support.Infrastructure.ZLink.Actors;
using SupportChat.Shared.Contracts;
using Zlink.Framework.Contracts.Handlers;
using Zlink.Framework.Contracts.Spots;

namespace SupportChat.Server.Support.Infrastructure.ZLink.Spots.EntrySpot.Handlers;

internal sealed class JoinConversationActorHandler
    : IZLinkEntrySpotActorSendHandler<
        SupportEntrySpot,
        SupportUserActor,
        JoinConversationReq>
{
    public ValueTask HandleAsync(
        SupportEntrySpot entrySpot,
        SupportUserActor actor,
        IZLinkMessageContext context,
        JoinConversationReq message,
        CancellationToken cancellationToken)
    {
        _ = entrySpot;
        _ = message;
        cancellationToken.ThrowIfCancellationRequested();
        if (!string.Equals(
                actor.Role,
                SupportChatRoles.Agent,
                StringComparison.Ordinal))
        {
            throw new InvalidOperationException(
                "Only agent conversation actors can join through this handler.");
        }

        var conversationId = context.Metadata.Find(
            SampleNames.ConversationIdMetadataKey)
            ?? throw new InvalidOperationException(
                "Conversation Join is missing the ConversationId metadata.");
        actor.TrackDeferredJoin(conversationId, notifyBoundSession: true);
        actor.Context.JoinSpot(
                conversationId,
                new JoinConversationReq(
                    actor.ParticipantId,
                    actor.Role,
                    actor.DisplayName))
            .Defer();
        return actor.Context.BoundSession
            .Send(new JoinConversationRes(
                true,
                new ConversationState(
                    conversationId,
                    string.Empty,
                    ConversationStatuses.WaitingForAgent,
                    string.Empty,
                    null,
                    0,
                    null,
                    null)))
            .Metadata(SampleNames.ConversationIdMetadataKey, conversationId)
            .Async(cancellationToken);
    }
}

File: framework/languages/java/samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/infrastructure/zlink/spots/entryspot/handlers/JoinConversationActorHandler.kt
package systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.spots.entryspot.handlers

import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpotActorRequestHandler
import systems.zlink.samples.kotlin.supportchat.server.configuration.SupportChatRoles
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.actors.SupportUserActor
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.spots.entryspot.SupportEntrySpot
import systems.zlink.samples.kotlin.supportchat.shared.contracts.JoinConversationReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.JoinConversationRes

class JoinConversationActorHandler :
    ZLinkSuspendingEntrySpotActorRequestHandler<
        SupportEntrySpot,
        SupportUserActor,
        JoinConversationReq,
        JoinConversationRes,
    > {
    override suspend fun handle(
        entrySpot: SupportEntrySpot,
        actor: SupportUserActor,
        context: ZLinkMessageContext,
        request: JoinConversationReq,
    ): JoinConversationRes {
        if (actor.role != SupportChatRoles.Agent) {
            throw IllegalStateException(
                "Only agent conversation actors can join through the Entry Spot"
            )
        }
        val conversationId =
            request.conversationId.takeIf(String::isNotBlank)
                ?: throw IllegalStateException("Conversation Join requires a conversation ID")
        return actor.scheduleConversationJoin(
            conversationId,
            "",
            JoinConversationReq(conversationId, actor.participantId, actor.role, actor.displayName),
        )
    }
}
